status: add project status

darkowlzz opened the pull request:
<!--
Work-in-progress PRs are welcome as a way to get early feedback - just prefix
the title with [WIP].

Add the change in the changelog (except for test changes and docs updates).
Please edit CHANGELOG.md and add the change under the appropriate category (NEW
FEATURES, IMPROVEMENTS & BUG FIXES) along with the PR number.
-->

### What does this do / why do we need it?
Adds `status` for project arguments.

`dep status github.com/foo/bar`

### What should your reviewer look out for in this PR?
The printed output formatting and logic of fetching all the project details.

### Do you need help or clarification on anything?
~Need to figure out ways to fetch some attributes like SOURCE TYPE, UPSTREAM EXISTS, UPSTREAM VERSION EXISTS using gps.~

<!--
### Which issue(s) does this PR fix?

fixes #
fixes #

-->

sdboyer commented on cmd/dep/status.go:
```diff
@@ -643,3 +648,140 @@ func collectConstraints(ptree pkgtree.PackageTree, p *dep.Project, sm gps.Source
+	Revision              string
+	LatestAllowed         string
+	SourceType            string
+	Packages              []string
```
The comment about `ProjectImports` and `PackageImports` made me realize that it would also be worth listing the ignores that are currently affecting this project - both what packages within it are ignored (that would otherwise actually be imported), and which of its imports are ignored.

This has the potential to be a little tricky, as if there's a hash mismatch (or if the user manually messed with `Gopkg.lock`), then it's possible that there may be some packages listed in `Gopkg.lock` that actually should be ignored. There's also a fair bit of work involved in actually pulling out the ignoreds, so i think it'll be fine for us to defer this to a later PR.

i've updated the spec with new lines and some comments explaining this, so we can discuss/dissect there; just noting it here.

sdboyer commented on cmd/dep/status.go:
```diff
@@ -643,3 +648,140 @@ func collectConstraints(ptree pkgtree.PackageTree, p *dep.Project, sm gps.Source
+	SourceType            string
+	Packages              []string
+	ProjectImporters      map[string]string
+	PackageImporters      map[string]string
```
i'd also be fine with leaving this to a follow-up PR - it's more important that we get something basically functional then ship additional bits of data later on.

darkowlzz replied:
I was working on adding `ProjectImports`, which I realized could be obtained from `collectConstraints()`, so I made the necessary changes in #962 . But yeah, we can add it as a follow-up PR.

sdboyer commented on cmd/dep/status.go:
```diff
@@ -643,3 +648,140 @@ func collectConstraints(ptree pkgtree.PackageTree, p *dep.Project, sm gps.Source
+
+		// Get the currently selected version from from lock.
+		for _, pl := range p.Lock.Projects() {
+			if pr == pl.Ident().ProjectRoot {
```
best practice here would be to minimize the indent level:

```go
if pr != pl.Ident().ProjectRoot {
	continue
}
```

then we can deindent the rest of the block, and the reader can immediately recognize that we're effectively just searching the slice for that single item.

sdboyer commented on cmd/dep/status.go:
```diff
@@ -643,3 +648,140 @@ func collectConstraints(ptree pkgtree.PackageTree, p *dep.Project, sm gps.Source
+			}
+		}
+
+		fmt.Println(projStatus)
```
print through the `ctx.Out` logger

sdboyer commented on cmd/dep/status.go:
```diff
@@ -643,3 +648,140 @@ func collectConstraints(ptree pkgtree.PackageTree, p *dep.Project, sm gps.Source
+
+				// Update local copy of the source and then fetch all the versions.
+				sm.SyncSourceFor(pl.Ident())
+				pvs, err := sm.ListVersions(pl.Ident())
```
in the current implementation, `ListVersions()` is already guaranteed to always return the latest set of versions, so the `SyncSourceFor()` call is unnecessary.

that may change when the caching work goes in, but we don't need to code defensively against that eventuality - we'll cross that bridge when we come to it.

sdboyer commented on cmd/dep/status.go:
```diff
@@ -643,3 +648,140 @@ func collectConstraints(ptree pkgtree.PackageTree, p *dep.Project, sm gps.Source
+
+				rev, _, _ := gps.VersionComponentStrings(pl.Version())
+				// REVISION
+				projStatus.Revision = rev
```
no need for the intermediate `rev` var, just assign directly to `projStatus.Revision`

darkowlzz commented:
Almost there. Figured out how to gather PACKAGE IMPORTS & PROJECT IMPORTS (hopefully correct 😅). LATEST ALLOWED is the only thing left. Once, we have all the data, I'll focus on the presentation.

darkowlzz commented:
Made some optimizations like advanced collection of package reachmaps and package list, to bring down the running time. Now, status of a single project or multiple projects take almost the same time (tried with 1 to 4 projects on dep itself, ~36s). It can be made concurrent to bring that down. Follow-up PR.

Sample output:
```
$ dep status github.com/Masterminds/vcs

PROJECT:			github.com/Masterminds/semver
VERSION:			parse-constraints-with-dash-in-pre
CONSTRAINTS:		[]
SOURCE:			github.com/Masterminds/semver
ALT SOURCE:		https://github.com/carolynvs/semver.git
PUB VERSION:		map[semver:[1.0.0 1.0.1 1.1.0 v1.1.1 v1.2.0 v1.2.1 v1.2.2 v1.2.3 v1.3.0 v1.3.1] branches:[2.x has-implied-caret master parse-constraints-with-dash-in-pre] nonsemvers:[]]
REVISION:		a93e51b5a57ef416dac8bb02d11407b6f55d8929
LATEST ALLOWED:		a93e51b5a57ef416dac8bb02d11407b6f55d8929
SOURCE TYPE:		git
PACKAGES:		github.com/Masterminds/semver
PROJECT IMPORTERS:	map[]
PACKAGE IMPORTERS:	map[]
UPSTREAM EXISTS:		true
UPSTREAM VERSION EXISTS:	true
```

PROJECT IMPORTERS and PACKAGE IMPORTERS are empty in this case because we don't have any dependency that uses semver project. An example for this for `github.com/dgryski/go-metro` would be:
```
PROJECT IMPORTERS:	map[github.com/jmank88/nuts:true]
PACKAGE IMPORTERS:	map[github.com/dgryski/go-metro:[github.com/jmank88/nuts/cmd/testpaths]]
```

CONSTRAINTS is empty for now because it depends on `constraintsCollection()`, which is implemented in #962 .

NEXT: Presentation improvements.

darkowlzz commented:
Improved the output using tabwriter and added an integration test.

I'll add the json output in a follow-up PR, avoiding huge change-set.

Example output:
```
$ dep status github.com/sdboyer/deptest

PROJECT:                  github.com/sdboyer/deptest
VERSION:                  v1.0.0
CONSTRAINTS:              []
SOURCE:                   github.com/sdboyer/deptest
ALT SOURCE:
PUB VERSION:              branches: master
                          nonsemvers:
                          semvers: v0.8.0, v0.8.1, v1.0.0
REVISION:                 ff2948a2ac8f538c4ecd55962e919d1e13e74baf
LATEST ALLOWED:           ff2948a2ac8f538c4ecd55962e919d1e13e74baf
SOURCE TYPE:              git
PACKAGES:                 github.com/sdboyer/deptest
PROJECT IMPORTERS:        github.com/darkowlzz/deptest-project-status, github.com/sdboyer/deptestdos
PACKAGE IMPORTERS:        github.com/sdboyer/deptest
                            github.com/darkowlzz/deptest-project-status/pkgbar
                            github.com/darkowlzz/deptest-project-status/pkgfoo
                            github.com/sdboyer/deptestdos
UPSTREAM EXISTS:          yes
UPSTREAM VERSION EXISTS:  yes
```

Leaving `CONSTRAINTS` as it is for now since its String() output would depend on the data structure, which would be changed in #962 .

darkowlzz commented:
Rebased and updated project status to print project constraints in proper table output.
```
CONSTRAINTS:              ^0.5.0(btb.com/x/y)
                          ^1.0.0(gh.com/f/b)

```

darkowlzz commented on cmd/dep/status.go:
```diff
@@ -827,10 +834,41 @@ func (pi packageImporters) TabString() string {
+	for _, pc := range pcs {
+		constList = append(constList, pc.String())
+	}
+	sort.Strings(constList)
```
This sorting would be simplified and improved with `projectConstraint` sort by Project implemented in #1413 .

darkowlzz commented:
Rebased. The end result for `CONSTRAINTS` field would be improved by #1452 which would collect root project constraints too.

mvdan commented:
Dep was officially deprecated earlier this year, and the proposal to archive this repository [was accepted](https://github.com/golang/go/issues/38158). As such, I'm closing outstanding issues before archiving the repository. For any further comments, please use the proposal thread on the Go issue tracker. Thanks!
